### PR TITLE
Adding exporting/importing capabilities

### PR DESCRIPTION
**Pull request by javuto**
This is a quite substantial diff so it will require testing and review. Also there are errors in ```hh_client```

### CODE REVIEW

File: src/controllers/exporter/ExporterController.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
+<?hh // strict
+
+abstract class ExporterController {
+  abstract public function genData(string $type): Awaitable<mixed>;
+
+  public async function genTeams(): Awaitable<mixed> {
+		$all_teams_data = array();
```
Comment (gsingh93) [inline src/controllers/exporter/ExporterController.php]: You're mixing tabs/spaces in this file, convert all the tabs to spaces.

File: src/controllers/exporter/ExporterController.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  	$all_teams = \HH\Asio\join(Team::genAllTeams());
```
Comment (gsingh93) [inline src/controllers/exporter/ExporterController.php]: Why not just use `await`?

File: src/controllers/exporter/JSONExporterController.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  public static function genJSON(mixed $data): string {	
+    return json_encode($data, JSON_PRETTY_PRINT);   
+  }
+
+  public static function sendJSON(mixed $data, string $json_file='fbctf.json'): void {
+    header('Content-Type: application/json;charset=utf-8');
+    header('Content-Disposition: attachment; filename='.$json_file);
+    print self::genJSON($data);
```
Comment (gsingh93) [inline src/controllers/exporter/JSONExporterController.php]: Let's stick with `echo` to be consistent.

File: src/controllers/importer/ImporterController.php [hunk 4]
```diff
@@ -0,0 +1,134 @@
+<?hh // strict
+
+abstract class ImporterController {
+  abstract public function importData(string $input_filename, string $type): ?array<string, mixed>;
+
+  public async function processTeams(array<string, mixed> $teams): Awaitable<bool> {
```
Comment (gsingh93) [inline src/controllers/importer/ImporterController.php]: Maybe the argument type should be `array<string, array<string, mixed>>`? That also might fix your hack errors. Same argument for the other functions.

File: src/controllers/importer/ImporterController.php [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    }
+    return true;
+  }
+
+  public async function importLogos(string $input_filename): Awaitable<bool> {
+    $logos = $this->importData($input_filename, 'logos');
+    return await $this->processLogos($logos);
+  }
+
+  public async function importGame(string $input_filename): Awaitable<bool> {
```
Comment (gsingh93) [inline src/controllers/importer/ImporterController.php]: One problem I see with the way you're doing this function is that it's possible for part of the import to succeed, which might be bad. i.e. we could successfully process logos, and then fail to import teams.

File: src/controllers/importer/JSONImporterController.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
+<?hh // strict
+
+class JSONImporterController extends ImporterController {
+	<<__Override>>
```
Comment (gsingh93) [inline src/controllers/importer/JSONImporterController.php]: Convert the tabs to spaces in this file.

File: src/models/Control.php [hunk 7]
```diff
@@ -112,22 +111,93 @@ class Control extends Model {
...
+    $game = await $exporter->genData('game');
+    JSONExporterController::sendJSON($game, 'fbctf_game.json');
+    exit();
+  }
+
+  public static async function exportTeams(): Awaitable<void> {
+    $exporter = new JSONExporterController();
+    $teams = await $exporter->genData('teams');
+    JSONExporterController::sendJSON($teams, 'fbctf_teams.json');
+    exit();
```
Comment (gsingh93) [inline src/models/Control.php]: Can you explain why we need these exit calls? We should just structure the code so they're not needed.

File: src/models/Token.php [hunk 8]
```diff
@@ -57,6 +57,7 @@ private static function generate(): string {
 
   // Create token.
   public static async function genCreate(): Awaitable<void> {
+    error_log('TOKEN!');
```
Comment (gsingh93) [inline src/models/Token.php]: remove

File: src/static/js/admin.js [hunk 9]
```diff
@@ -381,14 +381,113 @@ function databaseBackup() {
...
+  $('#import-levels_file').change(function() {
+    submitImport('levels_file', 'import_levels');
+  });
+}
+
+// Export and download current teams
+function exportCurrentTeams() {
+  var csrf_token = $('input[name=csrf_token]')[0].value;
+  var action = 'export_teams';
+  var url = 'index.php?p=admin&ajax=true&action=' + action + '&csrf_token=' + csrf_token;
```
Comment (gsingh93) [inline src/static/js/admin.js]: I don't like the idea of putting the csrf token in the URL. Can we make this a POST parameter?

**Comment by gsingh93** [comment]
This doesn't need to be done in this PR, but I think there's a cleaner way to do the data model to exported data conversion. We can have `Exportable`/`Importable` interfaces that classes like `Team` implements. Then when we want to make a new piece of data exportable/importable, we only have to implement that interface for that new piece of data.

**Comment by facebook-github-bot-4** [comment]
Thank you for your pull request.  As you may know, we require contributors to sign our Contributor License Agreement, and we don't seem to have you on file and listed as active anymore.  In order for us to review and merge your code, please email cla@fb.com with your details so we can update your status.

**Comment by facebook-github-bot-8** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Pull request closed without merging**